### PR TITLE
CODETOOLS-7902929: JMH generators fail for benchmark in unnamed package

### PR DESCRIPTION
**Pull request by retronym**
Prior to the patch, the enclosed test failed with:

```
$ mvn -Preflection test
...
Running UnnamedPackageTest
Tests run: 1, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0.004 sec <<< FAILURE! - in UnnamedPackageTest
compileTest(UnnamedPackageTest)  Time elapsed: 0.004 sec  <<< FAILURE!
java.lang.AssertionError: Annotation generator had thrown the exception.:
java.lang.NullPointerException
	at org.junit.Assert.fail(Assert.java:89)
	at org.junit.Assert.assertTrue(Assert.java:42)
	at UnnamedPackageTest.compileTest(UnnamedPackageTest.java:55)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
```

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Change must be properly reviewed

### Issue
 * [CODETOOLS-7902929](https://bugs.openjdk.java.net/browse/CODETOOLS-7902929): JMH generators fail for benchmark in unnamed package

### Reviewers
 * [Aleksey Shipilev](https://openjdk.java.net/census#shade) (@shipilev - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jmh pull/38/head:pull/38` \
`$ git checkout pull/38`

Update a local copy of the PR: \
`$ git checkout pull/38` \
`$ git pull https://git.openjdk.java.net/jmh pull/38/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 38`

View PR using the GUI difftool: \
`$ git pr show -t 38`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jmh/pull/38.diff">https://git.openjdk.java.net/jmh/pull/38.diff</a>

</details>

### CODE REVIEW

File: jmh-core-ct/src/test/java/UnnamedPackageTest.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    @Test
+    public void compileTest() {
+        InMemoryGeneratorDestination destination = new InMemoryGeneratorDestination();
+        boolean success = CompileTest.doTest(getClass(), destination);
+        if (success) {
+            Assert.fail("Should have failed.");
+        }
+        String error = destination.getErrors().get(0);
+        Assert.assertTrue(error, error.contains("Benchmark class should have package other than default."));
+    }
```
Comment (retronym) [inline jmh-core-ct/src/test/java/UnnamedPackageTest.java:56]: If you want to keep this test, I'm happy to refactor this, e.g. to take the expected message into an overload of `CompileTest.assertFail` .

File: jmh-core-ct/src/test/java/UnnamedPackageTest.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    @State(Scope.Benchmark)
+    public static class S {
+        @Setup(Level.Trial)
+        public void setup(AtomicInteger v) {}
+    }
+
+    @Benchmark
+    public void test(S s) {
+
+    }
```
Comment (shipilev) [inline jmh-core-ct/src/test/java/UnnamedPackageTest.java:45]: For the matter of this test, we can leave only:

```
@Benchmark
public void test() {}
```

...right?

**Review by shipilev — CHANGES_REQUESTED** [review]
In addition to patch comments, there are asm processor test failures with this new test, see GHA checks. It should be fixed e.g. like this:

```
diff --git a/jmh-generator-asm/src/main/java/org/openjdk/jmh/generators/asm/ASMClassInfo.java b/jmh-generator-asm/src/main/java/org/openjdk/jmh/generators/asm/ASMClassInfo.java
index f435d353..51b0442d 100644
--- a/jmh-generator-asm/src/main/java/org/openjdk/jmh/generators/asm/ASMClassInfo.java
+++ b/jmh-generator-asm/src/main/java/org/openjdk/jmh/generators/asm/ASMClassInfo.java
@@ -79,11 +79,24 @@ class ASMClassInfo extends ClassVisitor implements ClassInfo {
         this.superName = superName;
         this.idName = name;
         this.access = access;
-        qualifiedName = name.replace("/", ".");
-        packageName = qualifiedName.substring(0, qualifiedName.lastIndexOf("."));
-        origQualifiedName = qualifiedName;
-        qualifiedName = qualifiedName.replace('$', '.');
-        this.name = qualifiedName.substring(qualifiedName.lastIndexOf(".") + 1);
+        this.qualifiedName = name.replace("/", ".");
+
+        int dotIndex = qualifiedName.lastIndexOf(".");
+        if (dotIndex > 0) {
+            packageName = qualifiedName.substring(0, dotIndex);
+        } else {
+            packageName = "";
+        }
+
+        this.origQualifiedName = qualifiedName;
+        this.qualifiedName = qualifiedName.replace('$', '.');
+
+        dotIndex = qualifiedName.lastIndexOf(".");
+        if (dotIndex > 0) {
+            this.name = qualifiedName.substring(dotIndex + 1);
+        } else {
+            this.name = qualifiedName;
+        }
     }
 
     @Override
```

You can reproduce this locally by running `mvn ... -Pasm`.

File: jmh-core-ct/src/test/java/UnnamedPackageTest.java (hunk 2 above)
Comment (retronym) [reply jmh-core-ct/src/test/java/UnnamedPackageTest.java:45]: Yes! I was rushing here and left this copy/paste detritus.

File: jmh-generator-asm/src/main/java/org/openjdk/jmh/generators/asm/ASMClassInfo.java [hunk 3]
```diff
@@ -79,11 +79,14 @@ public void visit(int version, int access, String name, String signature, String
...
-        qualifiedName = qualifiedName.replace('$', '.');
-        this.name = qualifiedName.substring(qualifiedName.lastIndexOf(".") + 1);
+        this.qualifiedName = name.replace("/", ".");
+
+        int dotIndex = qualifiedName.lastIndexOf(".");
+        if (dotIndex != -1) {
+            packageName = qualifiedName.substring(0, dotIndex);
+        } else {
+            packageName = "";
+        }
```
Comment (shipilev) [inline jmh-generator-asm/src/main/java/org/openjdk/jmh/generators/asm/ASMClassInfo.java:89]: This is not the same as the previous code. Previous code handles "inner" classes with `$` additionally. See my previous comment how this should be resolved. (It is kinda open question if we really need this, but better to keep the semantics the same).

File: jmh-core-ct/src/test/java/org/openjdk/jmh/ct/CompileTest.java [hunk 4]
```diff
@@ -88,7 +88,7 @@ public static void assertOK(Class<?> klass) {
         }
     }
 
-    private static boolean doTest(Class<?> klass, InMemoryGeneratorDestination destination) {
+    public static boolean doTest(Class<?> klass, InMemoryGeneratorDestination destination) {
```
Comment (shipilev) [inline jmh-core-ct/src/test/java/org/openjdk/jmh/ct/CompileTest.java:91]: This visibility change is not required anymore.

**Review by shipilev — CHANGES_REQUESTED** [review]

File: jmh-generator-asm/src/main/java/org/openjdk/jmh/generators/asm/ASMClassInfo.java (hunk 3 above)
Comment (retronym) [reply jmh-generator-asm/src/main/java/org/openjdk/jmh/generators/asm/ASMClassInfo.java:89]: D'oh, now I can't even get the copy part of copy/paste right. Selected your entire patch and re-applied this time.

**Review by shipilev — APPROVED** [review]
Mondays, eh? Assuming tests are passing now, you are good to go!

**Comment by retronym** [comment]
Resubmitted as #39

**Pull request closed without merging**